File: storage_sdk/src/main/java/com/walker/storage/room/dao/UserDao.java
package com.walker.storage.room.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.walker.storage.room.model.User;
import com.walker.storage.room.relation.UserAndLibrary;
import com.walker.storage.room.relation.UserWithMusicLists;

import java.util.List;

/**
 * Created by walkerzpli on 2021/9/22.
 */
@Dao
public interface UserDao {

    @Insert
    void insertAll(User... users);

    @Delete
    void deleteById(User user);

    @Update
    void update(User... users);

    @Query("SELECT * FROM users WHERE id =:id")
    LiveData<User> getUser(int id);

    @Query("SELECT * FROM users")
    LiveData<List<User>> getAll();

    @Transaction
    @Query("SELECT * FROM users")
    LiveData<List<UserAndLibrary>> getAllUserWithLibrary();

    @Transaction
    @Query("SELECT * FROM users")
    LiveData<List<UserWithMusicLists>> getAllUsersWithMusicLists();

}

File: storage-ktx/src/main/java/com/demo/storage_ktx/dao/UserDao.kt
package com.demo.storage_ktx.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import com.demo.storage_ktx.model.User

/**
 * Created by walkerzpli on 2021/9/22.
 */
@Dao
interface UserDao {

    @Insert
    suspend fun insert(user: User)

    @Insert
    suspend fun insertAll(users: List<User>)

    @Query("DELETE FROM users WHERE first_name =:firstName")
    suspend fun delete(firstName: String)

    @Query("DELETE FROM users")
    suspend fun deleteAll()

    @Update
    suspend fun update(vararg users: User)

    @Query("SELECT * FROM users WHERE id =:id")
    fun getUser(id: Int): LiveData<User>

    @Query("SELECT * FROM users WHERE id =:id")
    fun getSingleUser(id: Int): User

    @Query("SELECT * FROM users")
    fun getAll(): LiveData<List<User>>

}